File: src/garry_types.c
/*
 * Copyright (c) 2026 Matthew Hughes (Argylelabcoat)
 * Licensed under the BSD-3-Clause License.
 * See LICENSE file for details.
 */

/**
 * @file garry_types.c
 * @brief Error code to string conversion for the Garry storage engine.
 *
 * Provides human-readable error messages for all status codes returned
 * by the storage engine API, enabling consistent error reporting.
 */

#include "garry/types.h"

/**
 * @brief Convert a status code to its string representation.
 *
 * @param status  The error code to look up.
 * @return Pointer to a static string describing the error.
 */
const char* garry_strerror(garry_status_t status)
{
    switch (status) {
        case GARRY_OK:                   return "ok";
        case GARRY_ERR_NOT_FOUND:        return "not found";
        case GARRY_ERR_LOCK_CONFLICT:    return "lock conflict";
        case GARRY_ERR_IO:               return "i/o error";
        case GARRY_ERR_CORRUPT:          return "corrupt database";
        case GARRY_ERR_INVALID_ARG:      return "invalid argument";
        case GARRY_ERR_NOMEM:            return "out of memory";
        case GARRY_ERR_BUFFER_TOO_SMALL: return "buffer too small";
        default:                         return "unknown error";
    }
}

File: wrappers/delphi/garry/garry_types.pas
unit garry_types;

{$IFDEF FPC}
  {$MODE DELPHI}
{$ENDIF}

interface

const
  GARRY_MAX_KEY_SIZE    = 256;
  GARRY_MAX_SUBSCRIPTS  = 16;
  GARRY_MAX_RECORD_SIZE = 16384;

type
  PGarryByte = ^Byte;
  PPAnsiChar = ^PAnsiChar;
  TGarryByteArray = array[0..GARRY_MAX_KEY_SIZE - 1] of Byte;

  PGarryDatabaseHandle = ^TGarryDatabaseHandle;
  TGarryDatabaseHandle = record
  end;

  PGarryCursorHandle = ^TGarryCursorHandle;
  TGarryCursorHandle = record
  end;

  TGarryStatus = (
    GarryOK                = 0,
    GarryErrNotFound       = 1,
    GarryErrLockConflict   = 2,
    GarryErrIO             = 3,
    GarryErrCorrupt        = 4,
    GarryErrInvalidArg     = 5,
    GarryErrNOMEM          = 6,
    GarryErrBufferTooSmall = 7
  );

  TGarryConfig = record
    PoolSize: Integer;
    MaxRecordSize: Integer;
    PageSize: Integer;
    MaxTxns: Integer;
    MaxVersions: Integer;
    MaxKeySize: Integer;
    MaxSubscripts: Integer;
    UseCompression: Boolean;
    BTreeFlags: Integer;
  end;

  TGarryKeyTuple = record
    Parts: array[0..GARRY_MAX_SUBSCRIPTS - 1] of PAnsiChar;
    Counts: array[0..GARRY_MAX_SUBSCRIPTS - 1] of Integer;
    Count: Integer;
  end;

  TGarryVisitor = procedure(const Key: PByte; KeyLen: Integer;
    const Value: PByte; ValueLen: Integer; UserData: Pointer); cdecl;

function GarryStatusToStr(Status: TGarryStatus): AnsiString;

implementation

function GarryStatusToStr(Status: TGarryStatus): AnsiString;
begin
  case Status of
    GarryOK:                Result := 'OK';
    GarryErrNotFound:       Result := 'not found';
    GarryErrLockConflict:   Result := 'lock conflict';
    GarryErrIO:             Result := 'I/O error';
    GarryErrCorrupt:        Result := 'corrupt';
    GarryErrInvalidArg:     Result := 'invalid argument';
    GarryErrNOMEM:          Result := 'out of memory';
    GarryErrBufferTooSmall: Result := 'buffer too small';
  else
    Result := 'unknown error';
  end;
end;

end.
